content/posts/lets-code-contact-form-in-phoenix-part-two.md

```diff
@@ -0,0 +1,240 @@
+
+Running the tests results in the following failure:
+
+```
+** (UndefinedFunctionError) function LearnU.Support.send_message/1 is undefined or private
+```
+
+I'm missing the `send_message` function. I can open up the `MyApp.Support` module again and add the following definition of the function to the bottom:
+
+```elixir
+# lib/my_app/support.ex
+
+defmodule MyApp.Support do
+  @doc """
+  Sends a message with given fields to the support team. Fields should include:
+  - `email`: Email for the user submitting the message for the support team to respond.
+  - `subject`: A short, general title as to the reason for support
+  - `body`: A longer description as to the reason for support and how the
+            support team might be able to help.
+
+  If the message is invalid, an `{:error, changeset}` is returned. Returns
+  `{:ok, message}` on success.
+  """
+  @spec send_message(map) :: {:ok, Message.t()} | {:error, Ecto.Changeset.t()}
+  def send_message(fields) when is_map(fields) do
+    %Message{}
+    |> Message.changeset(fields)
+    |> Ecto.Changeset.apply_action(:insert)
+  end
+end
+```
+
+Here, I have built a new `Message` changeset with the given `fields`, then passed that into [`Ecto.Changeset.apply_action/2`](https://hexdocs.pm/ecto/Ecto.Changeset.html#apply_action/2). I used `apply_action` here because the data is not being persisted to a database. This function will simply apply the changes and return the `:ok` tuple when the changes are valid. Otherwise, the function returns the `:error` tuple when the changes are invalid. Re-running the tests result in success.
+
+## Refactoring
+
+The eagle-eyed among you may have spotted an opportuntity for a refactor. The `send_message` and `change_message` functions both build a `Message` changeset. I can clean that up by simply allowing the `change_message` function to take an optional `map` of fields as an argument. I'm only going to show the changes here:
+
+```elixir
+# lib/my_app/support.ex
+
+defmodule MyApp.Support do
+  # I changed this function to take a `map` as an argument and set the default
+  # to an empty map.
+  @spec change_message(map) :: Ecto.Changeset.t()
+  def change_message(fields \\ %{}) do
+    %Message{}
+    |> Message.changeset(fields)
+  end
+
+  # This function was updated to use the new `change_message` function.
+  def send_message(fields) when is_map(fields) do
+    fields
+    |> change_message()
+    |> Ecto.Changeset.apply_action(:insert)
+  end
+end
+```
+
+Re-running the tests should result in success. Let's take a step back now and review the final product.
+
+## The Final Product
+
+```elixir
+# lib/my_app/support.ex
+
+defmodule LearnU.Support do
```

Comment:
Was the change from `MyApp` to `LearnU` intentional?

```suggestion
defmodule MyApp.Support do
```

Reply:
Good catch! The way I've been dogfooding this series is by implementing these features in another side-project. This is a simple copy-paste error from that project.